src/nodes/NodeConnectionManager.ts

```diff
@@ -0,0 +1,685 @@
+      clientFactory: async (args) =>
+        GRPCClientAgent.createGRPCClientAgent(args),
+    });
+    // Creating TTL timeout
+    const timer = setTimeout(async () => {
+      await this.destroyConnection(targetNodeId);
+    }, this.timeoutDelay);
+    // Add it to the map of active connections
+    const connectionAndLock = { connection, lock, timer };
+    this.connections.set(targetNodeId.toString(), connectionAndLock);
+    return connectionAndLock;
+  }
+
+  protected async destroyConnection(targetNodeId: NodeId): Promise<void> {
+    /**
+     * Get connAndLock from connections
+     * if connAndLock != null:
+     *   if conn != null:
+     *     acquire lock
+     *     destroy the conn
+     *     // if we destroy here, we need to ensure that any action using the conn must also
+     *     // acquire the lock, otherwise we could be performing an action whilst the conn
+     *     // is being destroyed
+     *     // we also need to consider a thread that may be waiting for the lock to be released
+     *     // if we destroy the conn and the lock, then what happens to the waiting thread?
+     *     delete from connections
+     */
+    const connAndLock = this.connections.get(targetNodeId.toString());
+    if (connAndLock == null) return;
+    const connection = connAndLock.connection;
+    if (connection == null) return;
+    const lock = connAndLock.lock;
+
+    // If the connection exists then we lock, destroy and remove it from the map
+    // @ts-ignore
+    await lock.withWrite(async () => {
+      // Destroying connection
+      await connection.destroy();
+      // Destroying TTL timer
+      if (connAndLock.timer != null) clearTimeout(connAndLock.timer);
+      // Updating the connection map
+      this.connections.set(targetNodeId.toString(), { lock });
+    });
+  }
+
+  // Migrated from NodeManager:
+  /**
+   * Treat this node as the server.
+   * Instruct the reverse proxy to send hole-punching packets back to the target's
+   * forward proxy, in order to open a connection from the client to this server.
+   * A connection is established if the client node's forward proxy is sending
+   * hole punching packets at the same time as this node (acting as the server)
+   * sends hole-punching packets back to the client's forward proxy.
+   * @param egressHost host of the client's forward proxy
+   * @param egressPort port of the client's forward proxy
+   * @param timer
+   */
+  @ready(new nodesErrors.ErrorNodeConnectionManagerNotRunning())
+  public async holePunchReverse(
+    egressHost: Host,
+    egressPort: Port,
+    timer?: Timer,
+  ): Promise<void> {
+    await this.revProxy.openConnection(egressHost, egressPort, timer);
+  } // Hole-punching
+
+  public async holePunchForward(
+    nodeId: NodeId,
+    ingressHost: Host,
+    ingressPort: Port,
+    timer?: Timer,
+  ): Promise<void> {
+    await this.fwdProxy.openConnection(nodeId, ingressHost, ingressPort, timer);
+  }
```

Comment:
This is only used in `NodeManager:pingNode`. I'd expect this to be used as part of the hole punch process when `holePunchReverse` is used but from the other side. I can't see that anywhere unless the forward connection is done automatically by the `ForwardProxy` when the `NodeConnection` is created and tries to connect? @joshuakarp any insights?

Reply:
Yeah, I'm pretty sure this whole function is completely unnecessary, because when we use the `ForwardProxy` to establish the connection within `NodeConnection`, it starts sending hole-punching packets to the target from the `ForwardProxy` internals (no need to call this function).

Reply:
What about when receiving a hole punch request message? That is on the receiving side you have to hole punch back?

Reply:
That's handled in a separate function and uses the `ReverseProxy`:

```ts
  // NodeConnectionManager
  @ready(new nodesErrors.ErrorNodeConnectionManagerNotRunning())
  public async holePunchReverse(
    egressHost: Host,
    egressPort: Port,
    timer?: Timer,
  ): Promise<void> {
    await this.revProxy.openConnection(egressHost, egressPort, timer);
  }
```

Reply:
Ok, I think that answers my question. So do I leave this as is or embed it into `NodeManager.pingNode()`?

Reply:
We do have the symmetry with having both `holePunchForward` and `holePunchReverse`, but given that we do the hole-punching with `ForwardProxy` internally anyway, perhaps it's best just to remove this to cause least confusion down the track?

Reply:
If this is being used by `pingNode` leave the method as is. Make sure to write docblock comment for both that explains how they are meant to be used, and that `holePunchForward` is not used when making regular connections because the `ForwardProxy` will automatically start hole punching upon a connection request.